src/datatypes/uncertainty.ts

```diff
@@ -55,41 +56,48 @@ export class Uncertainty {
   isPoint() {
     // Note: Can't use normal equality, as that fails for Javascript dates
     // TODO: Fix after we don't need to support Javascript date uncertainties anymore
-    const lte = (a: any, b: any) => {
-      if (typeof a !== typeof b) {
-        return false;
-      }
+    const lte = (a: any, b: any): boolean | null => {
+      if (typeof a !== typeof b || a?.constructor?.name !== b?.constructor?.name) return null;
+
       if (typeof a.sameOrBefore === 'function') {
         return a.sameOrBefore(b);
       } else {
         return a <= b;
       }
     };
-    const gte = (a: any, b: any) => {
-      if (typeof a !== typeof b) {
-        return false;
-      }
+    const gte = (a: any, b: any): boolean | null => {
+      if (typeof a !== typeof b || a?.constructor?.name !== b?.constructor?.name) return null;
+
       if (typeof a.sameOrBefore === 'function') {
         return a.sameOrAfter(b);
       } else {
         return a >= b;
       }
     };
     return (
-      this.low != null && this.high != null && lte(this.low, this.high) && gte(this.low, this.high)
+      this.low != null &&
+      this.high != null &&
+      (lte(this.low, this.high) ?? false) &&
+      (gte(this.low, this.high) ?? false)
     );
   }
 
   equals(other: any) {
+    // if this is a point, and other is not an uncertainty or a point, then we can compare directly
+    if (this.isPoint()) {
+      if (!(other instanceof Uncertainty)) return equals(this.low, other);
+
+      if (other instanceof Uncertainty && other.isPoint()) return equals(this.low, other.low);
+    }
+
     other = Uncertainty.from(other);
     return ThreeValuedLogic.not(ThreeValuedLogic.or(this.lessThan(other), this.greaterThan(other)));
   }
 
   lessThan(other: any) {
     const lt = (a: any, b: any) => {
-      if (typeof a !== typeof b) {
-        return false;
-      }
+      if (typeof a !== typeof b || a?.constructor?.name !== b?.constructor?.name) return null;
```

Comment:
This is nuanced.  So again, the functional composition I made above applies.  However, this time it makes a difference because there is no null coalescing operator.

I added a test for this to show the problem (Memoizer Tests/should properly match uncertainties line 125).  Suppose one uses `false`.

```
const u2 = new Uncertainty(new Date(2000, 1, 1));
const u7 = new Uncertainty(new DateTime(2000, 1, 1), new DateTime(2000, 1, 2));

// then equals(u2, u7) is true
```

Clearly they aren't equal (one is a point, and the other is not a point).

Here is debugging that case - notice that bestCase = worstCase = false

![image](https://user-images.githubusercontent.com/109760379/193694571-a7c7403d-401c-4b17-8c4b-0bdb1ec1b604.png)

That ultimately results in `ThreeValuedLogic.not(ThreeValuedLogic.or(this.lessThan(other), this.greaterThan(other)))` being true.

On the other hand, changing the return to `null` results in the following:

![image](https://user-images.githubusercontent.com/109760379/193695021-16bbdce6-0f29-4e19-9795-22fad50875ac.png)

So back to ThreeValuedLogic.  Comparing both:

- Using `false`: `ThreeValuedLogic.not(ThreeValuedLogic.or(false, false))` = `true`
- Using `null`: `ThreeValuedLogic.not(ThreeValuedLogic.or(null, null))` = `null`

So I changed to null to get a reasonable answer for this use case.  I did trace through the code to make sure the end users of equals would interpret `null` from equals as false and they do.

I believe this is a bug fix.

Reply:
Tricky. In a way, it is a bug fix-- but perhaps for a bug that should never surface anyway.  Since CQL [defines equal](https://cql.hl7.org/09-b-cqlreference.html#equal) as having balanced types (`=<T>(left T, right T) Boolean`), we should never get in the situation that your test case demonstrates.

If you do this in CQL:
```
define EqualWithDateAndDateTime: Date(2000, 1, 1) = DateTime(2000, 1, 1)
```
then the translator automatically does a conversion behind the scenes to convert the Date to a DateTime:
```json
"operand" : [ {
   "type" : "ToDateTime",
   "operand" : {
      "type" : "Date",
      "year" : { "valueType" : "{urn:hl7-org:elm-types:r1}Integer", "value" : "2000", "type" : "Literal" },
      "month" : { "valueType" : "{urn:hl7-org:elm-types:r1}Integer", "value" : "1", "type" : "Literal" },
      "day" : { "valueType" : "{urn:hl7-org:elm-types:r1}Integer", "value" : "1", "type" : "Literal" }
   }
}
```

If you compare two properties for which one of them is a Date and the other is a _choice_ between `Date` or `DateTime`, like this:
```
parameter DateOrDateTime Choice<Date, DateTime>
context Patient
define EqualWithDateAndDateOrDateTime: Date(2000, 1, 1) = DateOrDateTime
```
then the translator automatically casts the choice to a Date so the types match (and if the value is a DateTime, then the cast results in null):
```json
{
   "asType" : "{urn:hl7-org:elm-types:r1}Date",
   "type" : "As",
   "operand" : {
      "name" : "DateOrDateTime",
      "type" : "ParameterRef"
   }
}
```
If you make the constant a DateTime, then it just casts the choice to a DateTime instead.

If you try to just do a direct compare of two choices, the compiler yells at you:
> Equal(choice<System.Date,System.DateTime>,choice<System.Date,System.DateTime>) is ambiguous with:
>   - Equal(System.DateTime,System.DateTime)
>   - Equal(System.Date,System.Date)

So, anyway, I think maybe we didn't have a test for this because it's a comparison that isn't allowed by the spec. Still... no harm done in treating it properly if for some reason it ever got through -- so I'm fine w/ keeping it as you have it in this PR.